receiver/k8sclusterreceiver/internal/clusterresourcequota/clusterresourcequotas.go

```diff
@@ -0,0 +1,62 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package clusterresourcequota // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver/internal/clusterresourcequota"
+
+import (
+	"strings"
+	"time"
+
+	quotav1 "github.com/openshift/api/quota/v1"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver"
+	v1 "k8s.io/api/core/v1"
+
+	imetadataphase "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver/internal/clusterresourcequota/internal/metadata"
+)
+
+func GetMetrics(set receiver.CreateSettings, crq *quotav1.ClusterResourceQuota) pmetric.Metrics {
+	mbphase := imetadataphase.NewMetricsBuilder(imetadataphase.DefaultMetricsBuilderConfig(), set)
+	ts := pcommon.NewTimestampFromTime(time.Now())
+
+	emitCRQMetrics(mbphase, crq, crq.Status.Total.Hard, "openshift.clusterquota.limit", ts, "")
+	emitCRQMetrics(mbphase, crq, crq.Status.Total.Used, "openshift.clusterquota.used", ts, "")
+
+	for _, ns := range crq.Status.Namespaces {
+		emitCRQMetrics(mbphase, crq, ns.Status.Hard, "openshift.appliedclusterquota.limit", ts, ns.Namespace)
+		emitCRQMetrics(mbphase, crq, ns.Status.Hard, "openshift.appliedclusterquota.used", ts, ns.Namespace)
+
+	}
+
+	return mbphase.Emit()
+}
+
+func emitCRQMetrics(mbphase *imetadataphase.MetricsBuilder, crq *quotav1.ClusterResourceQuota, rl v1.ResourceList, metricName string, ts pcommon.Timestamp, namespace string) {
```

Comment:
That's fair. I figured it would be nice to avoid 4 almost identical for-loops with the method, but I see your point.

Reply:
I don't think you need 4 loops. it should be only one

Reply:
The loop is over the each of the `ClusterResourceQuota.ResourceList` objects. So we have to do the for loop over the resources governed by each of the given quota metrics.

`crq.Status.Total.Hard`, `crq.Status.Total.Used`, and then for each namespace within the cluster, do it namespace-by-namespace for `ns.Status.Hard` and `ns.Status.Used`.

Reply:
Just two nested loops don't work?

Reply:
No, because the resources are under the specific quota "unit" in the schema here. So the type hierarchy is:
The `ClusterResourceQuota.Status.Hard` property has a list of resources where the value is the hard limits for the given. The `ClusterResourceQuota.Status.Used` property has a list of resources that shows `current observed total usage of the resource`.

The for loop is then:
```
foreach resource governed by hard limit:
  emit hard limit quota metrics with this resource
```

If the `ClusterResourceQuota` [type](https://pkg.go.dev/github.com/openshift/api@v0.0.0-20230720094506-afcbe27aec7c/quota/v1#ClusterResourceQuota) had a top level resource list that then showed the values for different quotas on that resource, we could have a single loop, or something like nested loops.

Reply:
I've removed the extra method. After being able to use a single emit call, it doesn't looks as bad. Let me know if you'd prefer going back to the extra method though if this looks redundant.